pkg/net/libp2p/libp2ptransport.go

```diff
@@ -87,47 +86,50 @@ func (t *Transport) Stop() {
 			continue
 		}
 		t.logger.Log(logging.LevelDebug, "Closing connection", "to", id)
+
 		if err := s.Close(); err != nil {
-			t.logger.Log(logging.LevelWarn, fmt.Sprintf("Could not close connection to node %v: %v", id, err))
+			t.logger.Log(logging.LevelError, fmt.Sprintf("Could not close connection to node %v: %v", id, err))
 			continue
 		}
+
 		t.logger.Log(logging.LevelDebug, "Closed connection", "to", id)
 	}
 
+	t.host.RemoveStreamHandler(ID)
+
 	if err := t.host.Close(); err != nil {
 		t.logger.Log(logging.LevelError, fmt.Sprintf("Could not close libp2p %v: %v", t.ownID, err))
 	} else {
 		t.logger.Log(logging.LevelDebug, fmt.Sprintf("node %v libp2p host closed", t.ownID))
 	}
 }
 
-func (t *Transport) Connect(ctx context.Context) {
+func (t *Transport) Connect(ctx context.Context, nodes map[types.NodeID]types.NodeAddress) {
 	wg := sync.WaitGroup{}
-	wg.Add(len(t.membership) - 1)
+	if len(nodes) == 0 {
+		t.logger.Log(logging.LevelWarn, "failed to connect: empty map of nodes")
+		return
+	}
+	wg.Add(len(nodes) - 1)
```

Comment:
This takes a bit of brain gymnastics to read without any comments.
Could you either add a comment saying that`-1` is for the own connection, or even remove the `-1` and add a `wg.Done()` before line 117?
I think that would be easier to read.

Reply:
I agree. Fixed.